model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java

```diff
@@ -0,0 +1,365 @@
+import org.hibernate.annotations.TypeDef;
+import org.hibernate.annotations.TypeDefs;
+import org.keycloak.models.map.authorization.entity.MapPolicyEntity.AbstractMapPolicyEntity;
+import org.keycloak.models.map.common.DeepCloner;
+import org.keycloak.models.map.common.StringKeyConverter;
+import org.keycloak.models.map.common.UuidValidator;
+import org.keycloak.models.map.storage.jpa.JpaRootVersionedEntity;
+import org.keycloak.models.map.storage.jpa.hibernate.jsonb.JsonbType;
+
+import static org.keycloak.models.map.storage.jpa.Constants.CURRENT_SCHEMA_VERSION_AUTHZ;
+import org.keycloak.representations.idm.authorization.DecisionStrategy;
+import org.keycloak.representations.idm.authorization.Logic;
+
+/**
+ * There are some fields marked by {@code @Column(insertable = false, updatable = false)}.
+ * Those fields are automatically generated by database from json field,
+ * therefore marked as non-insertable and non-updatable to instruct hibernate.
+ */
+@Entity
+@Table(name = "kc_authz_policy", uniqueConstraints = {@UniqueConstraint(columnNames = {"realmId", "resourceServerId", "name"})})
+@TypeDefs({@TypeDef(name = "jsonb", typeClass = JsonbType.class)})
+public class JpaPolicyEntity extends AbstractMapPolicyEntity implements JpaRootVersionedEntity {
+
+    @Id
+    @Column
+    private UUID id;
+
+    //used for implicit optimistic locking
+    @Version
+    @Column
+    private int version;
+
+    @Type(type = "jsonb")
+    @Column(columnDefinition = "jsonb")
+    private final JpaPolicyMetadata metadata;
+
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
+    private Integer entityVersion;
+
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
+    private String realmId;
+
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
+    private UUID resourceServerId;
+
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
+    private String name;
+
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
+    private String owner;
+
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
+    private String type;
+
+    @Column(name = "scope_id")
+    @ElementCollection
+    @CollectionTable(name = "kc_authz_policy_scope", joinColumns = @JoinColumn(name = "policy_id", nullable = false))
+    private final Set<UUID> scopeIds = new HashSet<>();
+
+    @Column(name = "resource_id")
+    @ElementCollection
+    @CollectionTable(name = "kc_authz_policy_resource", joinColumns = @JoinColumn(name = "policy_id", nullable = false))
+    private final Set<UUID> resourceIds = new HashSet<>();
+
+    @Column(name = "associated_policy_id")
+    @ElementCollection
+    @CollectionTable(name = "kc_authz_policy_associated_policy", joinColumns = @JoinColumn(name = "policy_id", nullable = false))
+    private final Set<UUID> policyIds = new HashSet<>();
```

Comment:
Maybe, one more thing to consider would to replace `@ElementCollection`s here by `@OneToMany`. As I was reading about it more and it seems hibernate removes all elements and inserts all of them back with each update (addition or removal) [1].

@ahus1, @sguilhen wdyt? Should we leave it as is for now or do you prefer to possibly change it within this PR?

[1] https://thorben-janssen.com/hibernate-tips-elementcollection/

Reply:
@vramik - I'd be ok with this for now if the number of entries is small (in the range of 10-100) and would change seldom. 

Maybe one could add a comment to that.

Reply:
I've checked it with Pedro and it seems the number of objects should be probably between 10 and 30. I'll add a note for future then, thank you.